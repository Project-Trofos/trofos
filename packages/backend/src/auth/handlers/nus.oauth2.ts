import axios from "axios"
import jwt from "jsonwebtoken";

const HANDLER_NAME = 'NUS_OAUTH2';
const BASE_URL = 'https://luminus.nus.edu.sg/v2/api/login/adfstoken';
const GRANT_TYPE = "authorization_code";
const CLIENT_ID = "INC000002767827";
const RESOURCE = "sg_edu_nus_oauth";
const TOKEN_EMAIL_ADDRESS_KEY = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"

async function handleLogin(code: string, callbackUrl: string) {
    const data = new URLSearchParams();
    data.append('grant_type', GRANT_TYPE);
    data.append('client_id', CLIENT_ID);
    data.append('resource', RESOURCE);
    data.append('code', code);
    data.append('redirect_uri', callbackUrl);

    const config = {
        method: 'post',
        url: BASE_URL,
        data
      };
      

      const token = await axios(config);
      if (!token.data.access_token) {
        throw new Error("Error while fetching token from server");
      }

      const decodedToken = jwt.decode(token.data.access_token, {json : true});
      if (!decodedToken) {
        throw new Error("Error while decoding access token");
      }

      return decodedToken[TOKEN_EMAIL_ADDRESS_KEY];
}

export default {
    HANDLER_NAME,
    handleLogin
}