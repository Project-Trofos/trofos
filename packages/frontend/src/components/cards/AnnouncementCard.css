.announcement-card .ant-list-vertical .ant-list-item-meta-title {
  margin-bottom: 2px;
}

.announcement-card .announcement-card-content {
  white-space: pre-wrap;
}

.announcement-card .announcement-card-title-container {
  display: flex;
  justify-content: space-between;
}

.announcement-card .announcement-card-title-container .announcement-card-title-button-group {
  display: flex;
  gap: 5px;
}

.common-card .common-card-header .title {
  margin-bottom: 20px;
}
