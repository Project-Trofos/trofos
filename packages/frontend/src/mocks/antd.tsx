import React from 'react';
import { Select as AntdSelect, DatePicker as AntdDatePicker } from 'antd';
import mockDayjs from 'dayjs';

// Mock certain antd component to simplify testing in React Testing Library
vi.mock('antd', async () => { 
  const antd: any = await vi.importActual('antd');
  antd.theme.defaultConfig.hashed = false;

  function DatePicker(props: React.ComponentProps<typeof AntdDatePicker>) {
    const { defaultValue, placeholder, disabled, id, onChange } = props;
    return (
      <input
        id={id}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        placeholder={placeholder}
        disabled={disabled}
        onChange={(e) => {
          if (onChange) {
            onChange(mockDayjs(e.target.value) as unknown as Parameters<typeof onChange>['0'], e.target.value);
          }
        }}
      />
    );
  }

  function RangePicker(props: React.ComponentProps<typeof AntdDatePicker.RangePicker>) {
    const { defaultValue, id, onChange } = props;
    return (
      <input
        id={id as string | undefined}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLInputElement>['defaultValue']}
        onChange={(e) => {
          if (onChange) {
            onChange(
              [mockDayjs(e.target.value), mockDayjs(e.target.value)] as unknown as Parameters<typeof onChange>['0'],
              [e.target.value, e.target.value],
            );
          }
        }}
      />
    );
  }

  DatePicker.RangePicker = RangePicker;

  function Select(props: React.ComponentProps<typeof AntdSelect>) {
    const { mode, value, defaultValue, className, onChange, disabled, children, id } = props;

    const multiple = ['tags', 'multiple'].includes(mode ?? '');

    return (
      <select
        id={id}
        value={value as React.SelectHTMLAttributes<HTMLSelectElement>['value']}
        defaultValue={defaultValue as React.SelectHTMLAttributes<HTMLSelectElement>['defaultValue']}
        multiple={multiple}
        disabled={disabled}
        className={className}
        onChange={(e) => {
          if (onChange) {
            onChange(
              multiple ? Array.from(e.target.selectedOptions).map((option) => option.value) : e.target.value,
              {},
            );
          }
        }}
      >
        {children}
      </select>
    );
  }

  function Option({ children, ...otherProps }: { children: React.ReactNode }) {
    return <option {...otherProps}>{children}</option>;
  }
  function OptGroup({ children, ...otherProps }: { children: React.ReactNode }) {
    return <optgroup {...otherProps}>{children}</optgroup>;
  }

  Select.Option = Option;
  Select.OptGroup = OptGroup;

  return { ...antd, Select, DatePicker };
});
