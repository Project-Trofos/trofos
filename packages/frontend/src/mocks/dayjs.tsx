const dayjsMock = {
  year: () => '2022',
};

export default dayjsMock;
