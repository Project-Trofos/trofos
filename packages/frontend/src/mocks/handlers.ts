/* eslint-disable import/no-extraneous-dependencies */
import { rest } from 'msw';
import { CourseData } from '../api/types';

const BASE_URL = process.env.REACT_APP_BACKEND_BASE_URL || 'http://localhost:3001';
const NUSMODS_URL = 'https://api.nusmods.com/v2/2022-2023/moduleList.json';

export const MSW_COURSE: CourseData = {
  id: 1,
  code: 'CS3203',
  cname: 'Software Engineering Project',
  startYear: 2022,
  endYear: 2022,
  startSem: 1,
  endSem: 1,
  description: null,
  milestones: [
    {
      id: 1,
      name: 'Milestone 1',
      course_id: 1,
      start_date: '2023-12-31',
      deadline: '2023-01-09',
      created_at: '2022-12-18',
    },
    {
      id: 2,
      name: 'Milestone 2',
      course_id: 1,
      start_date: '2023-01-10',
      deadline: '2023-01-19',
      created_at: '2022-12-18',
    },
    {
      id: 3,
      name: 'Milestone 3',
      course_id: 1,
      start_date: '2023-01-20',
      deadline: '2023-01-30',
      created_at: '2022-12-18',
    },
  ],
  users: [],
  public: false,
  created_at: '2022-09-14T03:33:34.960Z',
};

const PROJECT = {
  id: 1,
  pname: 'project1',
  pkey: null,
  description: 'project1_description',
  course_id: '1',
  public: false,
  created_at: '2022-09-15T01:58:01.735Z',
  course: MSW_COURSE,
};

const NUSMODS_MODULE = {
  moduleCode: 'EL1101E',
  title: 'The Nature of Language',
  semesters: [1, 2],
};

const handlers = [
  // Handles GET on /project
  rest.get(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([PROJECT])))),

  // Handles GET on /course
  rest.get(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([MSW_COURSE])))),

  // Handles GET to NUSMODS
  rest.get(NUSMODS_URL, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify([NUSMODS_MODULE])))),

  // Handles POST on /project
  rest.post(`${BASE_URL}/project/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  // Handles POST on /course
  rest.post(`${BASE_URL}/course/project`, (req, res, ctx) => res(ctx.status(200), ctx.body(JSON.stringify({})))),

  rest.post(`${BASE_URL}/course/:courseId/milestone`, (req, res, ctx) =>
    res(ctx.status(200), ctx.body(JSON.stringify({}))),
  ),
];

export default handlers;
