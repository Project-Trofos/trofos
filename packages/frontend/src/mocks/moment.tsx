const momentMock = {
  year: () => '2022',
};

export default momentMock;
