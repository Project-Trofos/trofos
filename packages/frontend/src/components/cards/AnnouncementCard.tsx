import React from 'react';
import { Button, Card, List } from 'antd';
import { DeleteOutlined, UserOutlined } from '@ant-design/icons';

import { Subheading } from '../typography';
import { CourseData } from '../../api/types';
import AnnouncementCreationModal from '../modals/AnnouncementCreationModal';
import AvatarWithName from '../avatar/Avatar';
import Timestamp from '../common/Timestamp';
import AnnouncementUpdateModal from '../modals/AnnouncementUpdateModal';

import './CommonCard.css';
import './AnnouncementCard.css';

type AnnouncementCardProps = {
  course?: CourseData;
  showEdit?: boolean;
  handleDeleteAnnouncement: (announcementId: number) => Promise<void>;
  handleUpdateAnnouncement: (
    announcementId: number,
    payload: { announcementTitle?: string; announcementContent?: string },
  ) => Promise<void>;
};

export default function AnnouncementCard(props: AnnouncementCardProps): JSX.Element {
  const { course, showEdit = false, handleDeleteAnnouncement, handleUpdateAnnouncement } = props;

  return (
    <Card className="common-card announcement-card">
      <div className="common-card-header">
        <Subheading className="title">Announcements</Subheading>
        {course && showEdit && <AnnouncementCreationModal courseId={course.id.toString()} />}
      </div>
      <List
        itemLayout="vertical"
        bordered
        pagination={{ pageSize: 3 }}
        dataSource={course?.announcements}
        renderItem={(announcement) => (
          <List.Item key={announcement.id}>
            <List.Item.Meta
              title={
                <div className="announcement-card-title-container">
                  {announcement.title}
                  {showEdit && (
                    <div className="announcement-card-title-button-group">
                      <AnnouncementUpdateModal
                        initialTitle={announcement.title}
                        initialContent={announcement.content}
                        handleUpdate={(payload) => handleUpdateAnnouncement(announcement.id, payload)}
                      />
                      <Button size="small" danger onClick={() => handleDeleteAnnouncement(announcement.id)}>
                        <DeleteOutlined />
                      </Button>
                    </div>
                  )}
                </div>
              }
              description={
                <>
                  <Timestamp createdAt={announcement.created_at} updatedAt={announcement.updated_at} />
                  {announcement.user_id && (
                    <AvatarWithName
                      icon={<UserOutlined />}
                      // Will update to showing username after the API is available
                      username={course?.users.find((u) => u.user.user_id === announcement.user_id)?.user.user_email}
                    />
                  )}
                </>
              }
            />
            <section className="announcement-card-content">{announcement.content}</section>
          </List.Item>
        )}
      />
    </Card>
  );
}
