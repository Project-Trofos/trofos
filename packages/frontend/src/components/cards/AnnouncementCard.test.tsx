import React from 'react';
import { render, screen } from '@testing-library/react';
import AnnouncementCard from './AnnouncementCard';
import { MSW_COURSE } from '../../mocks/handlers';

describe('test AnnouncementCard', () => {
  describe('when rendering', () => {
    it('should render correctly', () => {
      const { baseElement } = render(
        <AnnouncementCard
          handleDeleteAnnouncement={jest.fn()}
          handleUpdateAnnouncement={jest.fn()}
          course={MSW_COURSE}
        />,
      );

      // Title and contents are shown
      for (const announcement of MSW_COURSE.announcements) {
        expect(screen.getByText(announcement.title)).toBeInTheDocument();
        expect(screen.getByText(announcement.content)).toBeInTheDocument();
      }

      // Compare with snapshot to ensure structure remains the same
      expect(baseElement).toMatchSnapshot();
    });
  });
});
