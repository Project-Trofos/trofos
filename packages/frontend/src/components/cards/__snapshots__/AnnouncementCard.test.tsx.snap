// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AnnouncementCard when rendering should render correctly 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered common-card announcement-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="common-card-header"
        >
          <h4
            class="ant-typography title"
          >
            Announcements
          </h4>
        </div>
        <div
          class="ant-list ant-list-vertical ant-list-split ant-list-bordered ant-list-something-after-last-item"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-list-item-no-flex"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <div
                          class="announcement-card-title-container"
                        >
                          ANNOUNCEMENT_TITLE
                        </div>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <div>
                          <div>
                            15/09/2022 11:33
                             
                            <span
                              class="ant-tag"
                            >
                              Edited
                            </span>
                          </div>
                        </div>
                        <div
                          class="avatar-container"
                        >
                          <span
                            class="ant-avatar ant-avatar-circle ant-avatar-icon"
                          >
                            <span
                              aria-label="user"
                              class="anticon anticon-user"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="user"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                                />
                              </svg>
                            </span>
                          </span>
                          <div />
                        </div>
                      </div>
                    </div>
                  </div>
                  <section
                    class="announcement-card-content"
                  >
                    ANNOUNCEMENT_CONTENT
                  </section>
                </li>
              </ul>
            </div>
          </div>
          <div
            class="ant-list-pagination"
          >
            <ul
              class="ant-pagination"
              unselectable="unselectable"
            >
              <li
                aria-disabled="true"
                class="ant-pagination-prev ant-pagination-disabled"
                title="Previous Page"
              >
                <button
                  class="ant-pagination-item-link"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    aria-label="left"
                    class="anticon anticon-left"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="left"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
              <li
                class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                tabindex="0"
                title="1"
              >
                <a
                  rel="nofollow"
                >
                  1
                </a>
              </li>
              <li
                aria-disabled="true"
                class="ant-pagination-next ant-pagination-disabled"
                title="Next Page"
              >
                <button
                  class="ant-pagination-item-link"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    aria-label="right"
                    class="anticon anticon-right"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="right"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
